app/lib/search/search_service.dart

```diff
@@ -418,6 +434,56 @@ class SearchQuery {
   }
 }
 
+/// Filter conditions on tags.
+class TagsPredicate {
+  /// tag -> required (true) | forbidden (false) map
+  final _values = <String, bool>{};
```

Comment:
nits, but two sets or arrays called `withTags` and `withoutTags`... or `requiredTags` and `negatedTags` might be easier...

If the same tags is in both we don't have to have special handling we just make it filter to nothing :)